dfget/core/api/uploader_api.go

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/dragonflyoss/Dragonfly/dfget/config"
+	"github.com/dragonflyoss/Dragonfly/dfget/util"
+)
+
+// UploaderAPI defines the communication methods between dfget and uploader.
+type UploaderAPI interface {
+	// Download downloads a piece and returns an HTTP response.
+	Download(req *DownloadRequest) (*http.Response, error)
+
+	// ParseRate sends a request to uploader to calculate the ratelimit dynamically
+	// for the speed limit of the whole host machine.
+	ParseRate(req *ParseRateRequest) (string, error)
+
+	// CheckServer checks the peer server on port whether is available.
+	CheckServer(req *CheckServerRequest) (string, error)
+
+	// FinishTask report a finished task to peer server.
+	FinishTask(req *FinishTaskRequest) error
+
+	// PingServer send a request to determine whether the server has started.
+	PingServer(req *CommonURL) bool
+}
+
+// uploaderAPI is an implementation of interface UploaderAPI.
+type uploaderAPI struct {
+}
+
+// NewUploaderAPI returns a new UploaderAPI.
+func NewUploaderAPI() UploaderAPI {
+	return &uploaderAPI{}
+}
+
+func (u *uploaderAPI) Download(req *DownloadRequest) (*http.Response, error) {
```

Comment:
I think that it's not a api of uploader server. It seems that dfget will download pieces only from uploader.

Reply:
Done.